File: src/utils/metrics.ts
/**
 * Performance Metrics Calculator
 * Calculate trading strategy performance metrics
 */

import { Trade, PerformanceMetrics } from '../types';

export class Metrics {
  /**
   * Calculate comprehensive performance metrics
   */
  static calculateMetrics(
    trades: Trade[],
    equity: { timestamp: number; value: number }[],
    initialCapital: number
  ): PerformanceMetrics {
    if (trades.length === 0 || equity.length === 0) {
      return this.getEmptyMetrics();
    }

    const finalCapital = equity[equity.length - 1].value;
    const totalReturn = (finalCapital - initialCapital) / initialCapital;

    // Calculate returns for each equity point
    const returns = equity.slice(1).map((point, i) =>
      (point.value - equity[i].value) / equity[i].value
    );

    // Separate winning and losing trades
    const { wins, losses } = this.separateTrades(trades);

    // Calculate time-based metrics
    const durationDays = (equity[equity.length - 1].timestamp - equity[0].timestamp) / (1000 * 60 * 60 * 24);
    const annualizedReturn = durationDays > 0 ?
      Math.pow(1 + totalReturn, 365 / durationDays) - 1 : 0;

    return {
      totalReturn: totalReturn * 100,
      annualizedReturn: annualizedReturn * 100,
      sharpeRatio: this.calculateSharpeRatio(returns, annualizedReturn),
      sortinoRatio: this.calculateSortinoRatio(returns, annualizedReturn),
      maxDrawdown: this.calculateMaxDrawdown(equity) * 100,
      winRate: this.calculateWinRate(trades) * 100,
      profitFactor: this.calculateProfitFactor(wins, losses),
      calmarRatio: this.calculateCalmarRatio(annualizedReturn, this.calculateMaxDrawdown(equity)),
      recoveryFactor: this.calculateRecoveryFactor(totalReturn, this.calculateMaxDrawdown(equity)),
      averageWin: this.calculateAverageWin(wins),
      averageLoss: this.calculateAverageLoss(losses),
      largestWin: this.calculateLargestWin(wins),
      largestLoss: this.calculateLargestLoss(losses),
      averageTradeDuration: this.calculateAverageTradeDuration(trades),
      expectancy: this.calculateExpectancy(wins, losses, trades.length),
    };
  }

  private static getEmptyMetrics(): PerformanceMetrics {
    return {
      totalReturn: 0,
      annualizedReturn: 0,
      sharpeRatio: 0,
      sortinoRatio: 0,
      maxDrawdown: 0,
      winRate: 0,
      profitFactor: 0,
      calmarRatio: 0,
      recoveryFactor: 0,
      averageWin: 0,
      averageLoss: 0,
      largestWin: 0,
      largestLoss: 0,
      averageTradeDuration: 0,
      expectancy: 0,
    };
  }

  private static separateTrades(trades: Trade[]): { wins: Trade[]; losses: Trade[] } {
    const wins: Trade[] = [];
    const losses: Trade[] = [];

    // Group trades into buy-sell pairs
    const pairs: [Trade, Trade][] = [];
    for (let i = 0; i < trades.length - 1; i += 2) {
      if (trades[i].side === 'buy' && trades[i + 1].side === 'sell') {
        pairs.push([trades[i], trades[i + 1]]);
      }
    }

    pairs.forEach(([buy, sell]) => {
      const pnl = (sell.price - buy.price) * buy.quantity - buy.fee - sell.fee;
      if (pnl > 0) {
        wins.push(sell);
      } else {
        losses.push(sell);
      }
    });

    return { wins, losses };
  }

  private static calculateWinRate(trades: Trade[]): number {
    if (trades.length < 2) return 0;

    const { wins } = this.separateTrades(trades);
    const totalPairs = Math.floor(trades.length / 2);

    return totalPairs > 0 ? wins.length / totalPairs : 0;
  }

  private static calculateProfitFactor(wins: Trade[], losses: Trade[]): number {
    if (losses.length === 0) return wins.length > 0 ? Infinity : 0;

    const grossProfit = wins.reduce((sum, trade) => sum + (trade.price * trade.quantity), 0);
    const grossLoss = losses.reduce((sum, trade) => sum + (trade.price * trade.quantity), 0);

    return grossLoss > 0 ? grossProfit / grossLoss : 0;
  }

  private static calculateSharpeRatio(returns: number[], annualizedReturn: number): number {
    if (returns.length === 0) return 0;

    const avgReturn = returns.reduce((a, b) => a + b, 0) / returns.length;
    const variance = returns.reduce((sum, r) => sum + Math.pow(r - avgReturn, 2), 0) / returns.length;
    const stdDev = Math.sqrt(variance);

    // Annualize the standard deviation
    const annualizedStdDev = stdDev * Math.sqrt(252); // 252 trading days

    return annualizedStdDev !== 0 ? annualizedReturn / annualizedStdDev : 0;
  }

  private static calculateSortinoRatio(returns: number[], annualizedReturn: number): number {
    if (returns.length === 0) return 0;

    const negativeReturns = returns.filter(r => r < 0);
    if (negativeReturns.length === 0) return annualizedReturn > 0 ? Infinity : 0;

    const avgNegReturn = negativeReturns.reduce((a, b) => a + b, 0) / negativeReturns.length;
    const downVariance = negativeReturns.reduce((sum, r) => sum + Math.pow(r - avgNegReturn, 2), 0) / negativeReturns.length;
    const downStdDev = Math.sqrt(downVariance);

    const annualizedDownStdDev = downStdDev * Math.sqrt(252);

    return annualizedDownStdDev !== 0 ? annualizedReturn / annualizedDownStdDev : 0;
  }

  private static calculateMaxDrawdown(equity: { timestamp: number; value: number }[]): number {
    let maxDrawdown = 0;
    let peak = equity[0].value;

    for (const point of equity) {
      if (point.value > peak) {
        peak = point.value;
      }

      const drawdown = (peak - point.value) / peak;
      maxDrawdown = Math.max(maxDrawdown, drawdown);
    }

    return maxDrawdown;
  }

  private static calculateCalmarRatio(annualizedReturn: number, maxDrawdown: number): number {
    return maxDrawdown !== 0 ? annualizedReturn / maxDrawdown : 0;
  }

  private static calculateRecoveryFactor(totalReturn: number, maxDrawdown: number): number {
    return maxDrawdown !== 0 ? totalReturn / maxDrawdown : 0;
  }

  private static calculateAverageWin(wins: Trade[]): number {
    if (wins.length === 0) return 0;
    return wins.reduce((sum, trade) => sum + (trade.price * trade.quantity - trade.fee), 0) / wins.length;
  }

  private static calculateAverageLoss(losses: Trade[]): number {
    if (losses.length === 0) return 0;
    return losses.reduce((sum, trade) => sum + (trade.price * trade.quantity + trade.fee), 0) / losses.length;
  }

  private static calculateLargestWin(wins: Trade[]): number {
    if (wins.length === 0) return 0;
    return Math.max(...wins.map(trade => trade.price * trade.quantity - trade.fee));
  }

  private static calculateLargestLoss(losses: Trade[]): number {
    if (losses.length === 0) return 0;
    return Math.min(...losses.map(trade => -(trade.price * trade.quantity + trade.fee)));
  }

  private static calculateAverageTradeDuration(trades: Trade[]): number {
    if (trades.length < 2) return 0;

    let totalDuration = 0;
    let count = 0;

    for (let i = 0; i < trades.length - 1; i += 2) {
      if (trades[i].side === 'buy' && trades[i + 1].side === 'sell') {
        totalDuration += (trades[i + 1].timestamp - trades[i].timestamp);
        count++;
      }
    }

    // Return in hours
    return count > 0 ? totalDuration / count / (1000 * 60 * 60) : 0;
  }

  private static calculateExpectancy(wins: Trade[], losses: Trade[], totalTrades: number): number {
    if (totalTrades === 0) return 0;

    const winRate = wins.length / totalTrades;
    const lossRate = losses.length / totalTrades;

    const avgWin = this.calculateAverageWin(wins);
    const avgLoss = Math.abs(this.calculateAverageLoss(losses));

    return (winRate * avgWin) - (lossRate * avgLoss);
  }

  /**
   * Format metrics for display
   */
  static formatMetrics(metrics: PerformanceMetrics): Record<string, string> {
    return {
      'Total Return': `${metrics.totalReturn.toFixed(2)}%`,
      'Annualized Return': `${metrics.annualizedReturn.toFixed(2)}%`,
      'Sharpe Ratio': metrics.sharpeRatio.toFixed(2),
      'Sortino Ratio': metrics.sortinoRatio.toFixed(2),
      'Max Drawdown': `${metrics.maxDrawdown.toFixed(2)}%`,
      'Win Rate': `${metrics.winRate.toFixed(2)}%`,
      'Profit Factor': metrics.profitFactor.toFixed(2),
      'Calmar Ratio': metrics.calmarRatio.toFixed(2),
      'Recovery Factor': metrics.recoveryFactor.toFixed(2),
      'Average Win': `$${metrics.averageWin.toFixed(2)}`,
      'Average Loss': `$${metrics.averageLoss.toFixed(2)}`,
      'Largest Win': `$${metrics.largestWin.toFixed(2)}`,
      'Largest Loss': `$${metrics.largestLoss.toFixed(2)}`,
      'Avg Trade Duration': `${metrics.averageTradeDuration.toFixed(2)}h`,
      'Expectancy': `$${metrics.expectancy.toFixed(2)}`,
    };
  }
}

File: src/openclaw_quant/metrics.py
"""
Performance Metrics Calculation

Calculate comprehensive trading performance metrics.
"""

import pandas as pd
import numpy as np
from typing import List, Dict
from .types import Trade


def calculate_metrics(
    trades: List[Trade],
    equity_curve: pd.Series,
    initial_capital: float
) -> Dict[str, float]:
    """
    Calculate comprehensive performance metrics.

    Args:
        trades: List of executed trades
        equity_curve: Equity over time
        initial_capital: Starting capital

    Returns:
        Dictionary of metrics
    """
    if len(trades) == 0 or len(equity_curve) == 0:
        return _empty_metrics()

    final_capital = equity_curve.iloc[-1]

    # Basic metrics
    total_return = final_capital - initial_capital
    total_return_pct = (total_return / initial_capital) * 100

    # Trade statistics
    buy_trades = [t for t in trades if t.side == 'buy']
    sell_trades = [t for t in trades if t.side == 'sell']

    # Match buy/sell pairs to calculate PnL
    trade_pnls = []
    for i in range(min(len(buy_trades), len(sell_trades))):
        buy = buy_trades[i]
        sell = sell_trades[i]
        pnl = (sell.price - buy.price) * buy.size - buy.fee - sell.fee
        trade_pnls.append(pnl)

    num_trades = len(trade_pnls)
    num_winning = sum(1 for pnl in trade_pnls if pnl > 0)
    num_losing = sum(1 for pnl in trade_pnls if pnl < 0)
    win_rate = (num_winning / num_trades * 100) if num_trades > 0 else 0.0

    # Win/Loss statistics
    winning_trades = [pnl for pnl in trade_pnls if pnl > 0]
    losing_trades = [pnl for pnl in trade_pnls if pnl < 0]

    avg_win = np.mean(winning_trades) if winning_trades else 0.0
    avg_loss = abs(np.mean(losing_trades)) if losing_trades else 0.0
    avg_trade = np.mean(trade_pnls) if trade_pnls else 0.0

    # Profit factor
    gross_profit = sum(winning_trades) if winning_trades else 0.0
    gross_loss = abs(sum(losing_trades)) if losing_trades else 0.0
    profit_factor = (gross_profit / gross_loss) if gross_loss > 0 else 0.0

    # Drawdown metrics
    max_drawdown, max_drawdown_pct = calculate_max_drawdown(equity_curve)

    # Risk-adjusted metrics
    sharpe_ratio = calculate_sharpe_ratio(equity_curve)
    sortino_ratio = calculate_sortino_ratio(equity_curve)

    return {
        'initial_capital': initial_capital,
        'final_capital': final_capital,
        'total_return': total_return,
        'total_return_pct': total_return_pct,
        'num_trades': num_trades,
        'num_winning': num_winning,
        'num_losing': num_losing,
        'win_rate': win_rate,
        'avg_trade': avg_trade,
        'avg_win': avg_win,
        'avg_loss': avg_loss,
        'profit_factor': profit_factor,
        'max_drawdown': max_drawdown,
        'max_drawdown_pct': max_drawdown_pct,
        'sharpe_ratio': sharpe_ratio,
        'sortino_ratio': sortino_ratio,
    }


def calculate_max_drawdown(equity: pd.Series) -> tuple[float, float]:
    """
    Calculate maximum drawdown.

    Returns:
        (absolute_drawdown, percentage_drawdown)
    """
    if len(equity) == 0:
        return 0.0, 0.0

    cummax = equity.cummax()
    drawdown = equity - cummax
    drawdown_pct = (drawdown / cummax) * 100

    max_dd = abs(drawdown.min())
    max_dd_pct = abs(drawdown_pct.min())

    return max_dd, max_dd_pct


def calculate_sharpe_ratio(equity: pd.Series, risk_free_rate: float = 0.0) -> float:
    """
    Calculate annualized Sharpe ratio.

    Args:
        equity: Equity curve
        risk_free_rate: Risk-free rate (annual)

    Returns:
        Sharpe ratio
    """
    if len(equity) < 2:
        return 0.0

    # Calculate returns
    returns = equity.pct_change().dropna()

    if len(returns) == 0 or returns.std() == 0:
        return 0.0

    # Annualize (assume daily data)
    mean_return = returns.mean() * 252  # Annualized
    std_return = returns.std() * np.sqrt(252)  # Annualized

    sharpe = (mean_return - risk_free_rate) / std_return

    return sharpe


def calculate_sortino_ratio(equity: pd.Series, risk_free_rate: float = 0.0) -> float:
    """
    Calculate annualized Sortino ratio (only penalize downside volatility).

    Args:
        equity: Equity curve
        risk_free_rate: Risk-free rate (annual)

    Returns:
        Sortino ratio
    """
    if len(equity) < 2:
        return 0.0

    # Calculate returns
    returns = equity.pct_change().dropna()

    if len(returns) == 0:
        return 0.0

    # Downside returns only
    downside_returns = returns[returns < 0]

    if len(downside_returns) == 0 or downside_returns.std() == 0:
        return 0.0

    # Annualize
    mean_return = returns.mean() * 252
    downside_std = downside_returns.std() * np.sqrt(252)

    sortino = (mean_return - risk_free_rate) / downside_std

    return sortino


def calculate_calmar_ratio(total_return: float, max_drawdown_pct: float) -> float:
    """
    Calculate Calmar ratio (return / max drawdown).

    Returns:
        Calmar ratio
    """
    if max_drawdown_pct == 0:
        return 0.0

    return total_return / max_drawdown_pct


def _empty_metrics() -> Dict[str, float]:
    """Return empty metrics dict"""
    return {
        'initial_capital': 0.0,
        'final_capital': 0.0,
        'total_return': 0.0,
        'total_return_pct': 0.0,
        'num_trades': 0,
        'num_winning': 0,
        'num_losing': 0,
        'win_rate': 0.0,
        'avg_trade': 0.0,
        'avg_win': 0.0,
        'avg_loss': 0.0,
        'profit_factor': 0.0,
        'max_drawdown': 0.0,
        'max_drawdown_pct': 0.0,
        'sharpe_ratio': 0.0,
        'sortino_ratio': 0.0,
    }


def format_metrics(metrics: Dict[str, float]) -> str:
    """Format metrics for display"""
    return f"""
Performance Metrics
==================
Total Return:     ${metrics['total_return']:,.2f} ({metrics['total_return_pct']:.2f}%)
Sharpe Ratio:     {metrics['sharpe_ratio']:.2f}
Sortino Ratio:    {metrics['sortino_ratio']:.2f}
Max Drawdown:     {metrics['max_drawdown_pct']:.2f}%

Trade Statistics
================
Total Trades:     {metrics['num_trades']}
Winning:          {metrics['num_winning']} ({metrics['win_rate']:.1f}%)
Losing:           {metrics['num_losing']}
Profit Factor:    {metrics['profit_factor']:.2f}
Avg Trade:        ${metrics['avg_trade']:.2f}
Avg Win:          ${metrics['avg_win']:.2f}
Avg Loss:         ${metrics['avg_loss']:.2f}
"""
